File: apps/study/e-buy-and-sell-crypto/basc.ts
import assert from 'node:assert'

function maxProfit(prices: number[]): number {
  let buy = 0
  let sell = 1
  let maxProfit = 0

  while (sell < prices.length) {
    if (prices[buy] < prices[sell]) {
      let profit = prices[sell] - prices[buy]
      if (profit > maxProfit) {
        maxProfit = profit
      }
    } else {
      buy = sell
    }
    sell++
  }

  return maxProfit
}

let target = 6
let prices = [10, 1, 5, 6, 7, 1]
assert.equal(maxProfit(prices), target)

target = 0
prices = [10, 8, 7, 5, 2]
assert.equal(maxProfit(prices), target)

File: apps/study/e-buy-and-sell-crypto/basc.go
package basc

import (
	"fmt"
	"math"
)

// func MaxProfit(prices []int) int {
// 	buy, sell := 0, 1
// 	maxProfit := 0

// 	for sell < len(prices) {
// 		if prices[buy] < prices[sell] {
// 			profit := prices[sell] - prices[buy]
// 			if profit > maxProfit {
// 				maxProfit = profit
// 			}
// 		} else {
// 			buy = sell
// 		}
// 		sell++
// 	}
// 	return maxProfit
// }

func MaxProfit(prices []int) int {
	maxP := 0
	minBuy := math.MaxInt32

	for _, sell := range prices {

		fmt.Printf("%d %d %d\n", sell, minBuy, maxP)

		if sell-minBuy > maxP {
			maxP = sell - minBuy
		}
		if sell < minBuy {
			minBuy = sell
		}
	}
	return maxP
}
